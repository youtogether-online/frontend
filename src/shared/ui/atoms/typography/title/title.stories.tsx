import type { Meta, StoryObj } from "@storybook/react";

import { Title } from "./title";

const meta: Meta<typeof Title> = {
  title: "Components/Typography/Title/Features  ",
  component: Title,
};

export default meta;
type Story = StoryObj<typeof Title>;

export const Orders: Story = {
  render: () => (
    <>
      <Title order={1}>H1</Title>
      <Title order={2}>H2</Title>
      <Title order={3}>H3</Title>
      <Title order={4}>H4</Title>
      <Title order={5}>H5</Title>
      <Title order={6}>H6</Title>
    </>
  ),
};
