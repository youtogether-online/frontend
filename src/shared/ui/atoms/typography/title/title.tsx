import { cva, type VariantProps } from "cva";
import { type ElementType, type ReactNode } from "react";

import { baseStyles } from "../styles";

type TitleComponentMapping = Record<string, ElementType>;

const titleComponentMapping: TitleComponentMapping = {
  1: "h1",
  2: "h2",
  3: "h3",
  4: "h4",
  5: "h5",
  6: "h6",
};

const title = cva("", {
  variants: {
    ...baseStyles,
    order: {
      1: ["text-5xl", "font-bold"],
      2: ["text-4xl", "font-bold"],
      3: ["text-3xl", "font-bold"],
      4: ["text-2xl", "font-bold"],
      5: ["text-xl", "font-semiBold"],
      6: ["text-lg", "font-semiBold"],
    },
  },
});

type TitleProps = VariantProps<typeof title> & {
  // Redefine `order` because cva doesn't provides way to set required props
  order: 1 | 2 | 3 | 4 | 5 | 6;
  children: ReactNode;
};

export const Title = ({ children, order, fontSize, fontWeight, mono }: TitleProps) => {
  const Component = titleComponentMapping[order];

  return <Component className={title({ order, fontSize, fontWeight, mono })}>{children}</Component>;
};
