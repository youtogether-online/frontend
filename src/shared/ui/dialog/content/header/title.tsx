import { DialogTitle } from "@radix-ui/react-dialog";
import { forwardRef, type ReactNode } from "react";

import { Title as TitleTypography } from "@/shared/ui/typography";

type TitleProps = {
  children: ReactNode;
};

export const Title = forwardRef<HTMLHeadingElement, TitleProps>(({ children }, ref) => {
  return (
    <DialogTitle asChild ref={ref}>
      <TitleTypography order={6}>{children}</TitleTypography>
    </DialogTitle>
  );
});
