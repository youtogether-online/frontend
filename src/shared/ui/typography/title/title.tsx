import { cva, type VariantProps } from "cva";
import { type ElementType, type ReactNode } from "react";
import { tw } from "typewind";

import { type SxProp } from "../../types";

type TitleComponentMapping = Record<string, ElementType>;

const titleComponentMapping: TitleComponentMapping = {
  1: "h1",
  2: "h2",
  3: "h3",
  4: "h4",
  5: "h5",
  6: "h6",
};

const title = cva("", {
  variants: {
    order: {
      1: [tw.text_5xl.font_bold],
      2: [tw.text_4xl.font_bold],
      3: [tw.text_3xl.font_bold],
      4: [tw.text_2xl.font_bold],
      5: [tw.text_xl.font_semiBold],
      6: [tw.text_lg.font_semiBold],
    },
  },
});

type TitleProps = VariantProps<typeof title> & {
  // Redefine `order` because cva doesn't provides way to set required props
  order: 1 | 2 | 3 | 4 | 5 | 6;
  children: ReactNode;
} & SxProp;

export const Title = ({ children, order }: TitleProps) => {
  const Component = titleComponentMapping[order];

  return <Component className={title({ order })}>{children}</Component>;
};
