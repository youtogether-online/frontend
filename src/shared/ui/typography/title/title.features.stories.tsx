import type { Meta, StoryObj } from "@storybook/react";

import { Title } from "./title";

const meta: Meta<typeof Title> = {
  title: "Components/Typography/Title/Features",
  tags: ["autodocs"],
  component: Title,
};

export default meta;
type Story = StoryObj<typeof Title>;

export const H1: Story = {
  args: {
    order: 1,
    children: "Heading 1",
  },
};

export const H2: Story = {
  args: {
    order: 2,
    children: "Heading 2",
  },
};

export const H3: Story = {
  args: {
    order: 3,
    children: "Heading 3",
  },
};

export const H4: Story = {
  args: {
    order: 4,
    children: "Heading 4",
  },
};

export const H5: Story = {
  args: {
    order: 5,
    children: "Heading 5",
  },
};

export const H6: Story = {
  args: {
    order: 6,
    children: "Heading 6",
  },
};
