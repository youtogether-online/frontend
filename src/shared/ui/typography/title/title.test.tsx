import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";

import { Title } from "./title";

describe("Title", () => {
  it("Title respects order prop", () => {
    const container = render(<Title order={1}>Heading 1</Title>);
    const title = container.getByRole("heading");
    expect(title).toMatchSnapshot();
  });
});
