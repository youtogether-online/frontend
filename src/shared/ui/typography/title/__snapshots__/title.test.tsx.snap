// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Title > Title respects order prop 1`] = `
<h1
  class="text-5xl font-bold"
>
  Heading 1
</h1>
`;
